precovery/utils.py

```diff
@@ -0,0 +1,136 @@
+import logging
+import multiprocessing as mp
+import signal
+
+import numpy as np
+import pandas as pd
+
+__all__ = ["yieldChunks", "calcChunkSize", "_initWorker", "_checkParallel"]
+
+logger = logging.getLogger(__name__)
+
+
+def yieldChunks(indexable, chunk_size):
+    """
+    Generator that yields chunks of size chunk_size.
+
+    Parameters
+    ----------
+    indexable : list, `~numpy.ndarray`, `~pandas.DataFrame`, `~pandas.Series` (N)
+        Indexable object that needs to be divided into
+        chunks.
+    chunk_size : int
+        Size of each chunk.
+
+    Yields
+    ------
+    chunk : indexable (<=chunk_size)
+        Chunks of indexable
+    """
+    if isinstance(indexable, list) or isinstance(indexable, np.ndarray):
+        for c in range(0, len(indexable), chunk_size):
+            yield indexable[c : c + chunk_size]
+    elif isinstance(indexable, pd.DataFrame) or isinstance(indexable, pd.Series):
+        for c in range(0, len(indexable), chunk_size):
+            yield indexable.iloc[c : c + chunk_size]
+    else:
+        err = "Indexable should be one of {list, `~numpy.ndarray`, `~pandas.DataFrame`, `~pandas.Series`}"
+        raise ValueError(err)
+
+
+def calcChunkSize(n, num_workers, max_chunk_size, min_chunk_size=1):
+    """
+    Calculate the optimal chunk size such that each worker gets at
+    least min_chunk_size chunks but does not get more than max_chunk_size
+    chunks. The goal is for no worker to be idle in terms of the number of items
+    it recieves
+
+    Parameters
+    ----------
+    n : int
+        Number of items.
+    num_workers : int
+        Number of workers to which items will be distributed.
+    max_chunk_size : int
+        Maximum chunk size to be given to each worker.
+    min_chunk_size : int, optional
+        Minimum chunk size to be given to each worker.
+
+    Yields
+    ------
+    chunk_size : int
+        Chunk_size between min_chunk_size and max_chunk_size.
+    """
+    # Calculate the number of n that should be sent to each worker
+    c = np.maximum(np.floor(n / num_workers), min_chunk_size).astype(int)
+
+    # Make sure this number does not exceed the maximum chunk size
+    chunk_size = np.minimum(c, max_chunk_size)
+    return chunk_size
+
+
+def _initWorker():
+    """
+    Tell multiprocessing worker to ignore signals, will only
+    listen to parent process.
+    """
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+    return
+
+
+def _checkParallel(num_jobs, parallel_backend):
```

Comment:
Good point - I think for this implementation in particular we can just stick with multiprocessing. I had taken out reference to the other backends in attribute_observations , but realize it was left in in this function. Will reorganize as you suggest.

This does have a difference with full-nelson - we really want single machines with many cores, because of the constraint in pulling data out of the indexed db. Doesn't scale horizontally as well for that reason, unless broken up in time.